File: scala/src/main/scala/0070_ClimbingStairs.scala
object ClimbingStairs extends App {
  /*

    n                               0
    1                    1.                    2.
    2                2.         3.      3.             4
    3            3      4     4.   5  4.   5         5    6

    brute force

    O(2^n), O(n)

    caching
    Arrary[stair] = nr of ways.
     [1,2,3]
   3 [2 1 1] 0

    O(n), O(n)


    n
    0. 3
    1. 2
    2. 1
    3  1

    O(n), O(1)
   */
  def climbStairs(n: Int): Int = {
    val s = Array(1, 1)

    for (i <- (n - 2) to 0 by -1) {
      s(i % 2) = s(i % 2) + s((i + 1) % 2)
    }

    s(0)
  }

  assert(climbStairs(2) == 2)
  assert(climbStairs(3) == 3)
}

File: ts/src/ts/0070_climbing_stairs.ts
import assertEqual from './assert';

/*
 * n = 3
 *
 * [0, 1, 2, 3]
 *  2  2  2  2 = 2^n
 *
 * brute force
 * O(2^n), O(n)
 *
 * dp: top down
 *
 * [0, 1, 2, 3]
 *  1  1  2  3
 *
 * O(n), O(2)
 */
function climbStairs(n: number): number {
  const dp: number[] = new Array(2).fill(0);

  dp[0] = 1;

  for (let i = 1; i <= n; i++) {
    dp[i % 2] += dp[(i + 1) % 2] as number;
  }

  return dp[n % 2] as number;
};

assertEqual(climbStairs(2), 2);
assertEqual(climbStairs(3), 3);



